### PR TITLE
Shuttle overhaul: this time with no areas

### PR DESCRIPTION
**Pull request by unid15**
This is a rework of #5202 which broke the game and had to be put down
Adresses #4998
https://github.com/Unfit/vgstation13/wiki/Guide-to-working-with-the-new-shuttle-system

---

This PR will make all shuttles use datums and docking ports instead of hardcode and areas. Yes, this means that there won't be any areas cluttering the map anymore and it'll be possible to build out of the escape airlock (but the emergency shuttle will still destroy everything)
It will also give admins a lot of commands to play with the new shuttle system, including creating custom shuttles

http://puu.sh/jb3Xv/5f21d2b97d.webm

Note that the shuttle itself is still considered an area, however the locations it can travel to aren't.
Shuttle rotation not supported yet

**Changelog**
- The following shuttles are now port based: MINING, RESEARCH, SYNDICATE, VOX SKIPJACK, EMERGENCY, TAXIS, CENTCOM FERRY, ARRIVALS SHUTTLE (can't be moved without adminbus)
- Shuttles now move some landmarks (latejoin, most importantly - so that the arrivals shuttle can be properly moved)
- Shuttles have dynamic lighting
- Added new shuttle control computers which can be linked to any shuttle
- Added admin commands that abuse the new shuttle system very hard
- New procs: shuttle_act(datum/shuttle/S) (for collision), can_shuttle_move(datum/shuttle/S) (for moving)

**Map changes:**
- Most of the areas belonging to changed shuttles have been removed and replaced with docking ports instead; the docking ports are invisible and don't affect gameplay at all
- Shuttles have been updated to include lights

**Other info:**
- There are two types of docking ports - shuttle and destination. Shuttle ports are installed on shuttles and connect to destination ports. Destination ports CAN be installed on shuttles, and, when moved by a shuttle they move the connected shuttle as well

**Milestones:**
22/07:
- Shuttles move properly

23/07:
- Ported most of the common shuttles (vox, syndicate, salvage, escape, mining, research)
- Updated boxstation
- Docking between shuttles
- Some adminbus commands

24/07:
- Finished all adminbus commands
- Made everything work properly
- Dynamic lighting
- Updated all maps

27/07:
- Removed oldcode

31/07:
- Pretty much ready for public testing

### CODE REVIEW

**Comment by PJB3005** [comment]
:+1:

**Comment by clusterfack** [comment]
FUCK AREAS. LETS DO THIS.

**Comment by clusterfack** [comment]
@probe1

File: code/game/area/areas.dm [hunk 1]
```diff
@@ -514,3 +514,326 @@
...
+		for(var/turf/simulated/T2 in fromupdate)
+			for(var/obj/machinery/door/D2 in T2)
+				doors += D2
+			/*if(T2.parent)
+				air_master.groups_to_rebuild += T2.parent
+			else
+				air_master.mark_for_update(T2)*/
+
+	for(var/obj/O in doors)
+		O:update_nearby_tiles()
```
Comment (clusterfack) [inline code/game/area/areas.dm]: Colon

File: code/game/area/areas.dm [hunk 2]
```diff
@@ -514,3 +514,326 @@
...
+
+						SX.air.copy_from(ST.zone.air)
+						ST.zone.remove(ST)
+
+					/* Quick visual fix for some weird shuttle corner artefacts when on transit space tiles */
+					if(direction && findtext(X.icon_state, "swall_s"))
+
+						// Spawn a new shuttle corner object
+						var/obj/corner = new()
+						corner.loc = X
```
Comment (PJB3005) [inline code/game/area/areas.dm]: ForceMove()

**Comment by Exxion** [comment]
Finally holy shit
I've always thought that areas were the worst way to do this, but it hadn't occurred to me that an alternative was feasible in BYOND

**Comment by Probe1** [comment]
@clusterfack d-does it mean we can use blueprints in the shuttle escape area to make new rooms...?

**Comment by clusterfack** [comment]
It does

**Comment by Probe1** [comment]
YES

**Comment by PJB3005** [comment]
There need to be explosions when a shuttle docks in an area that has turfs and machines and stuff inside it.

Atleast for dense objects.

**Comment by IratePirate** [comment]
Idea: Low chance of the shuttle being replaced with another shuttle type with a different layout, just to mix things up every so often.

**Comment by Unfit** [comment]
quite possible actually

(if somebody designs some alternative shuttles)

**Comment by IratePirate** [comment]
I could do that.

Is it possible while you're at it to fix corner shuttle wall sprites? They've been broken for like a year and make every shuttle look rather ugly.

**Comment by clusterfack** [comment]
My PR #5001 will most likely, fix shuttle corner walls and also allow automatic shuttle wall smoothing

**Comment by Unfit** [comment]
@IratePirate, post the designs here once you do them. Try to keep the shape and the size somewhat similar to the original ones, to prevent accidental collisions with the station
As for corners, I'll try

**Comment by Unfit** [comment]
oh god
@IratePirate
here's an idea

salvage shuttle
without spiders
but with BEARS

**Comment by Unfit** [comment]
this is pretty much done!
changelog and squash coming right up

File: code/datums/shuttle.dm [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+	.=..()
+	shuttles |= src
+
+	if(starting_area)
+		if(ispath(starting_area))
+			linked_area = locate(starting_area)
+		else if(isarea(starting_area))
+			linked_area = starting_area
+		else
+			world.log << "WARNING: Unable to find area [starting_area] in world - [src.type] ([src.name]) will be broken."
```
Comment (PJB3005) [inline code/datums/shuttle.dm]: You can use the proc `warning()` here to automatically put the `WARNING: ` and output it to `world.log`

File: html/changelogs/unfit_shuttles.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+# INDENT WITH TWO SPACES.  NOT TABS.  SPACES.
+# SCREW THIS UP AND IT WON'T WORK.
+# Also, this gets changed to [] after reading.  Just remove the brackets when you add new shit.
+changes: 
+  - rscadd: The following shuttles have been moved to a new shuttle system which doesn't use areas: mining shuttle, research shuttle, syndicate shuttle, emergency shuttle, vox shuttle, salvage shuttle
+  - rscdel: Areas of the shuttles above no longer clutter the map, and are replaced by invisible docking ports which do not affect gameplay in any way
+  - rscdel: Mining and research shuttle computers no longer work (and will soon be removed)
+  - rscadd: Added a new shuttle control console which can connect to any shuttle which allows it (by default - mining shuttle, research shuttle, salvage shuttle with password)
+  - experiment: Added dynamic lighting and lights to all shuttles
+  - experiment: Shuttles using the new shuttle system now throw all unanchored objects back when moving
```
Comment (Intigracy) [inline html/changelogs/unfit_shuttles.yml]: Does this only happen on initial movement?

Comment (Unfit) [inline html/changelogs/unfit_shuttles.yml]: It happens when they move (be it to another location or to/from transit)

Note that after testing it, I didn't really like its behaviour on syndicate/vox/salvage shuttles, which have tons of weapons and equipment lying around on the tables so the whole thing becomes a huge mess, so I might only enable it for mining, research and custom shuttles

**Comment by Unfit** [comment]
https://github.com/d3athrow/vgstation13/commit/6bc5b652416c5ae25c77bc47629ec242672bc468
POMF

**Comment by PJB3005** [comment]
POMF FUCKING WHY

**Comment by clusterfack** [comment]
My face when map priority fails

**Comment by MrStonedOne** [comment]
>66 commits
>none of them add shuttle rotation

Do you even shuttle?

**Comment by optimumtact** [comment]
At least they have non 
AREA SHUTTLES
R
E
A
S
H
U
T
T
L
E
S

**Comment by Probe1** [comment]
Merge to test server for testing.

**Comment by clusterfack** [comment]
How done/polished/tested is this? I would rather this go in sooner if possible to allow future edits to the relevant code here.

**Comment by Unfit** [comment]
Some minor issues with lighting (nothing gamebreaking through) and disappearing multi-tile airlocks
Give me a few hours to fix them

**Pull request closed without merging**